# Phase 2 — Data Understanding
Goal: Understand the structure, quality, and business meaning
of every column before any cleaning or analysis begins.

## Data Understanding Checklist
Before any analysis, a professional analyst checks:
1. Shape — rows and columns
2. Data types — are they correct
3. Missing values — which columns, how many
4. Duplicates — any repeated rows
5. Distributions — do values look realistic

This is called Exploratory Data Quality Assessment.
It happens before EDA. You cannot trust analysis
built on data you haven't verified.

In [ ]:
# ── PHASE 2: DATA UNDERSTANDING ──────────────────────────────
# Goal: Inspect the dataset before touching anything
# Analyst mindset: trust nothing until you verify it yourself
import pandas as pd
import numpy as np
# Load the raw data — always from the raw folder
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print("="*50)
print("STEP 1 - SHAPE")
print("="*50)
print(f"Rows    :{df.shape[0]}")
print(f"Columns : {df.shape[1]}") 

STEP 1 - SHAPE
Rows    :7043
Columns : 21


In [5]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')

print("="*50)
print("STEP 3 - MISSING VALUES")
print("="*50)
# .isnull() checks every single cell — True if empty, False if not
# .sum() counts how many True values exist per column
# This tells us exactly which columns have missing data and how many
missing = df.isnull().sum()
# We only want to see columns that actually have missing values
# so we filter: only show columns where missing count is above 0
print(missing[missing>0])
print(f"\nTotal missing values in entire dataset : {df.isnull().sum()}")

STEP 3 - MISSING VALUES
Series([], dtype: int64)

Total missing values in entire dataset : customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print("="*50)
print("STEP 4 - DUPLICATE ROWS")
print("="*50)
# .duplicated() checks every row — True if that row is an exact
# copy of a previous row
# .sum() counts how many duplicate rows exist
duplicates=df.duplicated().sum()
print(f"Duplicate rows found : {duplicates}")

STEP 4 - DUPLICATE ROWS
Duplicate rows found : 0


In [3]:
import pandas as pd
import numpy as np
print("="*50)
print("STEP 2 - DATA TYPES")
print("=")
# .dtypes shows us what type of data each column contains
# object = text, int64 = whole numbers, float64 = decimal numbers
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print(df.dtypes)

STEP 2 - DATA TYPES
=
customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object


In [9]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print("="*50)
print("=" * 50)
print("STEP 3B — HIDDEN MISSING VALUES IN TotalCharges")
print("=" * 50)

# df['TotalCharges'] selects just the TotalCharges column
# .str accesses string methods on that column
# .strip() removes any spaces from the start and end of each value
# == '' checks if after stripping spaces, the value is empty
# This catches values that look empty but are actually spaces

hidden = df[df['TotalCharges'].str.strip() == '']

print(f"Rows with hidden blank TotalCharges : {len(hidden)}")

STEP 3B — HIDDEN MISSING VALUES IN TotalCharges
Rows with hidden blank TotalCharges : 11


In [10]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
# Now let's check for values that look empty but aren't null
# .str.strip() removes spaces from text
# == '' checks if what remains is an empty string
# We do this on TotalCharges because it should be numeric
# but pandas loaded it as text — that's suspicious

spaces = df[df['TotalCharges'].str.strip() == '']
print(f"Rows where TotalCharges is blank string : {len(spaces)}")
print("\nWhat do these rows look like?")
print(spaces[['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges', 'Churn']].head(10))

Rows where TotalCharges is blank string : 11

What do these rows look like?
      customerID  tenure  MonthlyCharges TotalCharges Churn
488   4472-LVYGI       0           52.55                 No
753   3115-CZMZD       0           20.25                 No
936   5709-LVOEQ       0           80.85                 No
1082  4367-NUYAO       0           25.75                 No
1340  1371-DWPAZ       0           56.05                 No
3331  7644-OMVMY       0           19.85                 No
3826  3213-VVOLG       0           25.35                 No
4380  2520-SGTTA       0           20.00                 No
5218  2923-ARZLG       0           19.70                 No
6670  4075-WKNIU       0           73.35                 No


In [11]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
# Let's see what these problem rows actually look like
# We select only 5 columns to keep the output readable
# This helps us understand WHO these customers are

print(hidden[['customerID','tenure','MonthlyCharges',
              'TotalCharges','Churn']].to_string())

      customerID  tenure  MonthlyCharges TotalCharges Churn
488   4472-LVYGI       0           52.55                 No
753   3115-CZMZD       0           20.25                 No
936   5709-LVOEQ       0           80.85                 No
1082  4367-NUYAO       0           25.75                 No
1340  1371-DWPAZ       0           56.05                 No
3331  7644-OMVMY       0           19.85                 No
3826  3213-VVOLG       0           25.35                 No
4380  2520-SGTTA       0           20.00                 No
5218  2923-ARZLG       0           19.70                 No
6670  4075-WKNIU       0           73.35                 No
6754  2775-SEFEE       0           61.90                 No


In [ ]:

print("=" * 50)
print("STEP 3B — HIDDEN MISSING VALUES")
print("=" * 50)

# First let's look at TotalCharges specifically
# A customer's total charges should always be a number
# Let's check what data type it was assigned
print(f"TotalCharges data type : {df['TotalCharges'].dtype}")

# Now let's see the first 10 values of TotalCharges
# .head(10) shows the first 10 rows of a column
print("\nFirst 10 TotalCharges values:")
print(df['TotalCharges'].head(10))

STEP 3B — HIDDEN MISSING VALUES
TotalCharges data type : str

First 10 TotalCharges values:
0      29.85
1     1889.5
2     108.15
3    1840.75
4     151.65
5      820.5
6     1949.4
7      301.9
8    3046.05
9    3487.95
Name: TotalCharges, dtype: str


## Data Quality Finding — TotalCharges

Found 11 rows where TotalCharges is blank.
These are NOT data entry errors.
These customers have tenure = 0, meaning they signed up
but have not yet completed their first billing cycle.

Business meaning: these are day-zero customers.
TotalCharges is blank because no bill has been issued yet.

Decision: We will handle these in the cleaning phase.
Options:
- Fill TotalCharges with 0 (they have been charged nothing)
- Remove these 11 rows (too new to analyse for churn)
We will decide which option during cleaning based on
what the analysis requires.

In [13]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print("=" * 50)
print("STEP 5A — TARGET VARIABLE: CHURN")
print("=" * 50)

# .value_counts() counts how many times each unique value appears
# This tells us how many customers churned vs stayed
churn_counts = df['Churn'].value_counts()
print(churn_counts)

# Now calculate the percentage
# This tells us what proportion of customers are churning
print("\nChurn rate as percentage:")

# len(df) gives total number of rows
# We divide each count by total rows and multiply by 100
# .round(2) rounds to 2 decimal places
print((churn_counts / len(df) * 100).round(2))

STEP 5A — TARGET VARIABLE: CHURN
Churn
No     5174
Yes    1869
Name: count, dtype: int64

Churn rate as percentage:
Churn
No     73.46
Yes    26.54
Name: count, dtype: float64


In [14]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print("=" * 50)
print("STEP 5B — NUMERICAL COLUMN DISTRIBUTIONS")
print("=" * 50)

# .describe() gives us a statistical summary of every number column
# It shows: count, mean, min, max, and quartiles
# This tells us if values look realistic for a real business

# We select only the three numeric columns
numeric_cols = ['tenure', 'MonthlyCharges', 'TotalCharges']

# We can't run describe on TotalCharges yet because it's stored
# as text — so we convert it temporarily just for this view
# pd.to_numeric converts text to numbers
# errors='coerce' means: if a value can't be converted, make it NaN
df_temp = df.copy()
df_temp['TotalCharges'] = pd.to_numeric(
    df_temp['TotalCharges'], errors='coerce'
)

print(df_temp[numeric_cols].describe().round(2))

STEP 5B — NUMERICAL COLUMN DISTRIBUTIONS
        tenure  MonthlyCharges  TotalCharges
count  7043.00         7043.00       7032.00
mean     32.37           64.76       2283.30
std      24.56           30.09       2266.77
min       0.00           18.25         18.80
25%       9.00           35.50        401.45
50%      29.00           70.35       1397.48
75%      55.00           89.85       3794.74
max      72.00          118.75       8684.80


## Key Data Understanding Findings

### Churn Rate: 26.54%
Roughly 1 in 4 customers is leaving. This is the core
business problem we are solving.

### Tenure Distribution
- 25% of customers leave within 9 months
- The first 9 months = highest risk window
- Retention strategy must prioritise new customers

### Revenue Distribution
- Average customer generates $2,283 lifetime value
- Std deviation nearly equals mean — massive variation
- Budget vs premium customers behave very differently
- Customer segmentation is critical for this business

### Data Quality Issue Found
- TotalCharges stored as text, not number — must fix
- 11 rows have blank TotalCharges — tenure = 0 customers
- These will be handled in the cleaning phase

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print("=" * 50)
print("BUSINESS IMPACT OF 26.54% CHURN RATE")
print("=" * 50)

# Total customers in the dataset
total_customers = 7043

# Average monthly payment per customer
avg_monthly_charge = 64.76

# Total monthly revenue if every customer stays
total_monthly_revenue = total_customers * avg_monthly_charge

# Number of customers churning (26.54% of total)
# We convert percentage to decimal by dividing by 100
churned_customers = total_customers * (26.54 / 100)

# Monthly revenue lost to churn
# Each churned customer stops paying avg_monthly_charge
revenue_lost_monthly = churned_customers * avg_monthly_charge

# Annual revenue lost
# Multiply monthly loss by 12 months
revenue_lost_annually = revenue_lost_monthly * 12

print(f"Total customers          : {total_customers:,}")
print(f"Monthly revenue          : ${total_monthly_revenue:,.2f}")
print(f"Customers lost to churn  : {churned_customers:,.0f}")
print(f"Monthly revenue lost     : ${revenue_lost_monthly:,.2f}")
print(f"Annual revenue lost      : ${revenue_lost_annually:,.2f}")

BUSINESS IMPACT OF 26.54% CHURN RATE
Total customers          : 7,043
Monthly revenue          : $456,104.68
Customers lost to churn  : 1,869
Monthly revenue lost     : $121,050.18
Annual revenue lost      : $1,452,602.18


## Business Impact Analysis

Current annual churn cost: $1,452,602

This means every 10% improvement in churn retention
saves the business approximately $145,260 per year.

This is the financial justification for building a
churn prediction and intelligence system.

A data analyst who can quantify business impact in
dollars is 10x more valuable than one who only
produces charts.